File: src/main/java/com/anod/appwatcher/sync/ManualSyncService.java
package com.anod.appwatcher.sync;

import android.app.IntentService;
import android.content.ContentProviderClient;
import android.content.Intent;
import android.content.Context;
import android.os.Bundle;

import com.anod.appwatcher.AppListContentProvider;

/**
 * An {@link IntentService} subclass for handling asynchronous task requests in
 * a service on a separate handler thread.
 * <p>
 */
public class ManualSyncService extends IntentService {
    private static final String ACTION_MANUAL_SYNC = "com.anod.appwatcher.sync.action.SYNC";

    public ManualSyncService() {
        super("ManualSyncService");
    }

    /**
     * Starts this service to perform action Sync. If
     * the service is already performing a task this action will be queued.
     *
     * @see IntentService
     */
    public static void startActionSync(Context context) {
        Intent intent = new Intent(context, ManualSyncService.class);
        intent.setAction(ACTION_MANUAL_SYNC);
        context.startService(intent);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        if (intent != null) {
            SyncAdapter syncAdapter = new SyncAdapter(getApplicationContext());
            ContentProviderClient contentProviderClient = getContentResolver().acquireContentProviderClient(AppListContentProvider.AUTHORITY);

            Bundle bundle = new Bundle();
            bundle.putBoolean(SyncAdapter.SYNC_EXTRAS_MANUAL, true);

            int updatesCount = syncAdapter.onPerformSync(bundle, contentProviderClient);

            Intent finishIntent = new Intent(SyncAdapter.SYNC_STOP);
            finishIntent.putExtra(SyncAdapter.EXTRA_UPDATES_COUNT, updatesCount);
            sendBroadcast(finishIntent);
        }
    }

}

File: src/main/java/com/anod/appwatcher/sync/ManualSyncService.kt
package com.anod.appwatcher.sync

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.work.workDataOf
import com.anod.appwatcher.BuildConfig
import info.anodsplace.framework.app.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * An [IntentService] subclass for handling asynchronous task requests in
 * a service on a separate handler thread.
 *
 *
 */
class ManualSyncService : IntentService("ManualSyncService") {

    private val coroutineScope = CoroutineScope(Dispatchers.Default)

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            val syncAdapter = UpdateCheck(ApplicationContext(applicationContext))

            val data = workDataOf(
                UpdateCheck.extrasManual to !BuildConfig.DEBUG
            )
            coroutineScope.launch {
                val updatesCount = syncAdapter.perform(data)
                val finishIntent = Intent(UpdateCheck.syncStop)
                finishIntent.putExtra(UpdateCheck.extrasUpdatesCount, updatesCount)
                sendBroadcast(finishIntent)
            }
        }
    }

    companion object {
        private const val ACTION_MANUAL_SYNC = "com.anod.appwatcher.sync.action.SYNC"

        /**
         * Starts this service to perform action Sync. If
         * the service is already performing a task this action will be queued.

         * @see IntentService
         */
        fun startActionSync(context: Context) {
            val intent = Intent(context, ManualSyncService::class.java)
            intent.action = ACTION_MANUAL_SYNC
            context.startService(intent)
        }
    }

}
